### PR TITLE
8322619: Parts of SG no longer update during rendering - overlapping - culling - dirty

### PR DESCRIPTION
**Pull request by FlorianKirmaier**
In some situations, a part of the SG is no longer rendered.
I created a test program that showcases this problem.

Explanation:

This can happen, when a part of the SG, is covered by another Node.
In this part, one node is totally covered, and the other node is visible.

When the totally covered Node is changed, then it is marked dirty and it's parent, recursively until an already dirty node is found.
Due to the Culling, this totally covered Node is not rendered - with the effect that the tree is never marked as Clean.

In this state, a Node is Dirty but not It's parent. Based on my CodeReview, this is an invalid state which should never happen.

In this invalid state, when the other Node is changed, which is visible, then the dirty state is no longer propagated upwards - because the recursive "NGNode.markTreeDirty" algorithm encounters a dirty node early.

This has the effect, that any SG changes in the visible Node are no longer rendered. Sometimes the situation repairs itself.

Useful parameters for further investigations:
-Djavafx.pulseLogger=true
-Dprism.printrendergraph=true
-Djavafx.pulseLogger.threshold=0

PR:
This PR ensures the dirty flag is set to false of the tree when the culling is used.
It doesn't seem to break any existing tests - but I'm not sure whether this is the right way to fix it.
It would be great to have some feedback on this solution - maybe guiding me to a better solution.

I could write a test, that just does the same thing as the test application, but checks every frame that these nodes are not dirty - but maybe there is a better way to test this.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8322619](https://bugs.openjdk.org/browse/JDK-8322619): Parts of SG no longer update during rendering - overlapping - culling - dirty (**Bug** - P4) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).
 * [JDK-8332040](https://bugs.openjdk.org/browse/JDK-8332040): Parts of SG no longer update during rendering - overlapping - culling - dirty (**CSR**) (Withdrawn)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1310/head:pull/1310` \
`$ git checkout pull/1310`

Update a local copy of the PR: \
`$ git checkout pull/1310` \
`$ git pull https://git.openjdk.org/jfx.git pull/1310/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1310`

View PR using the GUI difftool: \
`$ git pr show -t 1310`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1310.diff">https://git.openjdk.org/jfx/pull/1310.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1310#issuecomment-1866177932)

### CODE REVIEW

**Comment by FlorianKirmaier** [comment]
*Fixed wrong ticket-number in title and commit*

**Review by kevinrushforth — COMMENTED** [review]
I'd like to see an automated (probably robot-based) test, if it is feasible.

This will need careful review. 

Reviewers: @arapte @lukostyra

**Comment by FlorianKirmaier** [comment]
If someone who understands the rendering could provide some feedback on what is happening - than I can probably make a better fix, and write a reasonable unit-test for this bug.
The current fix is a bit suboptimal - because the call to clean-dirty is O(n) and might be called n time - leading to O(n^2)

**Comment by hjohn** [comment]
The fully covered node which was dirty is now marked clean in the PR when it is culled.  Now that it is clean, what happens when it is uncovered (being careful not to change it in any other way)?  Does it get rendered correctly, or does it miss the last change made to it while it was fully covered?

If that works correctly still, then I think this is the right fix.  Culling the node is (IMHO) the same as rendering it (there is nothing to render) so it should be clean after being culled.

**Comment by FlorianKirmaier** [comment]
It is worth noting that the idea of cleaning the dirty tree, for the culled node, is also done at line `NGNode:1414`, but for some reason, it doesn't catch the case from this bug.

**Comment by FlorianKirmaier** [comment]
I tried out a bit, by removing the clipping area in my test program.
I added the changed program to the ticket.
Didn't notice anything wrong.
There are probably many ways to test for regressions. 

Can't guarantee that there are no other cases that go wrong - but if something doesn't work, then it happens more rarely than this one.

File: .idea/jpa-buddy.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (FlorianKirmaier) [reply .idea/jpa-buddy.xml]: Removed the accidental change!

**Comment by lukostyra** [comment]
> Maybe we could make snapshot based tests? That should work on all platforms, right?

We have a whole suite of tests which use similar methods to perform checks, called robot tests. Since you want to check if a node was properly updated, and the best way to do that seems to be checking RGB values, I think a test for this case belongs right there.

You could use robot utility to check color values of specific pixels on the screen. I think the best way to go about it would be to do something similar to what your test bug apps do (the ones attached to the JBS issue). Ideally I would try and look for first some form of confidence check (if correct controls are "on" and "off"), then progressing the animation and seeing if controls updated as intended.

All robot tests are located in `tests/system/src/test/java/test/robot` and then they follow package names. I'm not 100% sure where this test should go though (I assume `.../javafx/scene`?).

**Comment by eduardsdv** [comment]
I analysed the bug and can confirm the fix.
I summarized it in the following table. 
First column is the node structure, the others the steps: update nodes, render, update nodes, render, ...

| Nodes (L - left and R - right) / Steps			| 1st -> update Circles 	| 2nd -> render 	| 3rd -> update Lines 	| 4th -> render |
|-|-|-|-|-|
|HBox					|dirty				|clearDirty		|dirty				|clearDirty			
|HBox >StackPane-L			|dirty				|clearDirty		|dirty				|clearDirty
|HBox >StackPane-L > Group-L			|dirty				|clearDirty		|dirty				|clearDirty
|HBox >StackPane-L > Group-L >Line-L		|					|				|dirty				|clearDirty
|HBox >StackPane-L > Group-L >Circle-L		|dirty				|clearDirty	    	|
|HBox >StackPane-R			|dirty				|clearDirty  *1		|				|	*3
|HBox >StackPane-R > Group-R			|dirty				|dirty			|dirty  
|HBox >StackPane-R > Group-R > Line-R		|					|				|dirty  *2				|dirty
|HBox >StackPane-R > Group-R > Circle-R		|dirty				|dirty			|dirty				|dirty


*1: > no culling bits for first dirty region (Circle-L) -> skip children -> clearDirty() is not called on children
     > no root path for the second dirty region (Circle-R) -> no rendering and root.clearDirtyTree() stops on StackPane-R (the children stay dirty: Group-R and Circle-R)
	
*2: > Line-R.markTreeDirty() skips marking of further parents because Group-R is already dirty -> the StackPane-R stays clean

*3: > the dirty regions are not collected because the StackPane-R is clean -> Line-R is not rendered and the dirty-flag is not reset

-------
The error is that ``clearDirty()`` is not called in the second step on the children of StackPane-R (see *1). Subsequent changes in Line-R and Circle-R are not propagated to the root node because Group-R is already dirty (see *2) and therefore StackPane-R remains clean. The last changes are therefore ignored in the next rendering phase (see *3).

-------
We can simply say: **the dirty flags must be deleted on all nodes after rendering**.

If this is true, the test can simply check this instead of screenshots or similar.

**Comment by eduardsdv** [comment]
After further investigation and testing I would suggest to remove all ``clearDirty()`` and ``clearDirtyTree()`` calls and put only one clear-dirty pass after rendering is done (at the end of ``ViewerPainter.paintImpl(final Graphics backBufferGraphics)``).

Since the ``markDirty()`` method does both, sets the dirty flag and propagates it to the ancestor nodes, it is better if there is only one method for deleting the dirty flag from the node and all its children. Therefore, I would remove the ``clearDirtyTree()`` method and move its content to the ``clearDirty()`` method.

This way we can guarantee that all dirty flags are removed after the rendering is completed.
We also guarantee that a clean parent with dirty children will never occur (the bug this PR addresses).

Furthermore, my quick test shows that fewer ``clearDirty()`` calls are required in the new version. Bonus: We are even faster.

````java
    public void clearDirty() {
        if (dirty != DirtyFlag.CLEAN || childDirty) {
            dirty = DirtyFlag.CLEAN;
            childDirty = false;
            dirtyBounds.makeEmpty();
            dirtyChildrenAccumulated = 0;

            if (this instanceof NGGroup) {
                List<NGNode> children = ((NGGroup) this).getChildren();
                for (NGNode child : children) {
                    child.clearDirtyTree_();
                }
            }
        }

        if (getClipNode() != null) {
            getClipNode().clearDirty();
        }
    }

````

**Comment by eduardsdv** [comment]
I added a test that shows that the @FlorianKirmaier's fix works. You can start it with: `gradlew -PFULL_TEST=true :systemTests:test --tests NGNodeDirtyFlagTest`.

To avoid such errors in the future, I would still suggest the refactoring, which I wrote about in my last comment.

@kevinrushforth and @arapte please review.

**Comment by eduardsdv** [comment]
I created an alternative solution for this bug. See PR: https://github.com/openjdk/jfx/pull/1451

**Comment by FlorianKirmaier** [comment]
Sorry for creating the CSR, it was an accident.

Now that Eduard has both created an alternative solution, but also has reviewed that this solution is correct, and provided an unit-test - I think this (or the other) PR is ready to move forwards.

@kevinrushforth 
What do you think?

This is still quite a fundamental bug which is probably quite prevalent, 
because when it happens it is really hard to notice it.
But it probably happens more often than we think.

**Comment by kevinrushforth** [comment]
> Sorry for creating the CSR, it was an accident.

No problem.

> Now that Eduard has both created an alternative solution, but also has reviewed that this solution is correct, and provided an unit-test - I think this (or the other) PR is ready to move forwards.
> 
> @kevinrushforth What do you think?

Yes, I've asked @arapte to be the primary reviewer on this. I'll leave it to his judgment as to which of the two approaches to take forward.

**Comment by eduardsdv** [comment]
keep alive

**Comment by eduardsdv** [comment]
@arapte Please review.

**Comment by kevinrushforth** [comment]
@FlorianKirmaier since it looks like we are going with PR #1451 as the solution, please close this PR (or move it to Draft until #1451 is integrated and then close it).

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
This PR should be closed now that PR #1451 is integrated.

**Pull request closed without merging**